### PR TITLE
initial commit for pull request

### PR DESCRIPTION
**Pull request by dmccann3**

### CODE REVIEW

**Comment by ScottaBout** [comment]
Hi Dennis, I'll review your code.

File: dqn.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+import itertools
+import torch.nn as nn
+import torch.optim as optim
+import matplotlib.pyplot as plt
+import matplotlib.animation as animation
+
+from collections import namedtuple, deque
+from discreteaction_pendulum import Pendulum
+
+torch.manual_seed(0)
```
Comment (ScottaBout) [inline dqn.py:13]: Good idea for reproducibility.

File: dqn.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        out = self.layer2(out)
+        out = self.relu(out)
+        return self.layer3(out)
+    
+    def act(self, obs):
+        obs_t = torch.as_tensor(obs, dtype=torch.float32)
+        q_values = self(obs_t.unsqueeze(0))
+
+        max_q_index = torch.argmax(q_values, dim=1)[0]
+        action = max_q_index.detach().item()
```
Comment (ScottaBout) [inline dqn.py:56]: why do you need to detach()

File: dqn.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        sample = random.random()
+        if sample <= e:
+            a = random.randrange(self.env.num_actions)
+            return a
+        else:
+            a = self.Q.act(s)
+            return a 
+
+
+    def get_return(self, reward_list):
```
Comment (ScottaBout) [inline dqn.py:133]: Not used?

File: dqn.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        # compute targets
+        target_q_vals = self.target_Q(next_state_batch_t)
+        max_target_q_vals = target_q_vals.max(dim=1, keepdim=True)[0]
+        targets = reward_batch_t + GAMMA * (1 - dones_batch_t) * max_target_q_vals
+
+        # get action q vals
+        q_vals = self.Q(state_batch_t)
+        action_q_vals = torch.gather(input=q_vals, dim=1, index=action_batch_t)
+
+        criterion = nn.SmoothL1Loss()
```
Comment (ScottaBout) [inline dqn.py:167]: Have you found more success using L1 loss vs MSE loss?

File: dqn.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    def train(self):
+        self.fill_buffer(MIN_BUFFER_SIZE)
+
+        s = self.env.reset()
+        epi_reward = 0.0
+        for step in itertools.count():
+            a = self.get_action(s, step)
+
+            s1, r, done = self.env.step(a)
+            self.buffer.add( s, a, s1, r, done)
```
Comment (ScottaBout) [inline dqn.py:203]: Why do you have to fill your buffer with random values before adding the sampled values?

**Review by ScottaBout — COMMENTED** [review]
At initial look, the algorithm appears to be structured correctly. It is difficult to ensure all the dimensions are matched properly but it looks like it runs. Other than a couple questions, I do not see anything major needed to change. Well done!
